clients/application-kafka/src/main/java/org/eclipse/hono/application/client/kafka/impl/KafkaBasedCommandSender.java

```diff
@@ -124,8 +187,72 @@ public KafkaBasedCommandSender(
         Objects.requireNonNull(deviceId);
         Objects.requireNonNull(command);
 
-        // TODO to implement.
-        return null;
+        final Span span = TracingHelper
+                .buildChildSpan(tracer, context, "send command and wait for response", getClass().getSimpleName())
+                .withTag(Tags.SPAN_KIND.getKey(), Tags.SPAN_KIND_CLIENT)
+                .withTag(TracingHelper.TAG_TENANT_ID, tenantId)
+                .withTag(TracingHelper.TAG_DEVICE_ID, deviceId)
+                .start();
+        final String correlationId = correlationIdSupplier.get();
+        final Promise<Void> requestTracker = Promise.promise();
+        final Promise<DownstreamMessage<KafkaMessageContext>> responseTracker = Promise.promise();
+        final Promise<Void> timeOutTracker = Promise.promise();
+        final long timerId = vertx.setTimer(COMMAND_TIMEOUT, id -> {
+            timeOutTracker.fail(new SendMessageTimeoutException("sending command/waiting for response timed out after "
+                    + COMMAND_TIMEOUT + "ms"));
```

Comment:
I think the correlation id should be removed here from the `commandResponseSubscriptions.get(tenantId)` value list.

Reply:
I have incorporated this and the other comments and pushed the changes.